doc/rate-limiting.md

```diff
@@ -0,0 +1,340 @@
+
+An example to illustrate
+
+```yaml
+configurations:
+  - actions:
+    - request_headers:
+        header_name: "X-MY-CUSTOM-HEADER"
+        descriptor_key: "custom-header"
+        skip_if_absent: true
+  - actions:
+    - generic_key:
+        descriptor_key: admin
+        descriptor_value: "1"
+```
+
+A request without "X-MY-CUSTOM-HEADER" would generate one descriptor with one entry:
+
+```
+("admin": "1")
+```
+
+A request with a header "X-MY-CUSTOM-HEADER=MY-VALUE" would generate two descriptors,
+one entry each descriptor:
+
+```
+("admin": "1")
+("custom-header": "MY-VALUE")
+```
+
+**Note**: If one action is not able to populate a descriptor entry, the entire descriptor is discarded.
+
+**Note**: The external rate limiting service will be called only when there is at least one not empty
+descriptor.
+
+### Rate limiting configuration rules
+
+Configuration rules allow rate limit configurations to be activated conditionally depending on
+the current context (the incoming HTTP request properties).
+Each rate limit configuration list can define, optionally, a list of rules to match the request.
+A match occurs when *at least* one rule matches the request.
+
+An example to illustrate. Given these rate limit configurations,
+
+```yaml
+spec:
+    rateLimits:
+    - configurations:
+      - actions:
+        - generic_key:
+            descriptor_key: toystore-app
+            descriptor_value: "1"
+    - rules:
+      - hosts: ["api.toystore.com"]
+      configurations:
+      - actions:
+        - generic_key:
+            descriptor_key: api
+            descriptor_value: "1"
+    - rules:
+      - hosts: ["admin.toystore.com"]
+      configurations:
+      - actions:
+        - generic_key:
+            descriptor_key: admin
+            descriptor_value: "1"
+```
+
+* When a request for `api.toystore.com` hits the gateway, the descriptors generated would be:
+
+```
+("toystore-app", "1")
+("api", "1")
+```
+
+* When a request for `admin.toystore.com` hits the gateway, the descriptors generated would be:
+
+```
+("toystore-app", "1")
+("admin", "1")
+```
+
+* When a request for `other.toystore.com` hits the gateway, the descriptors generated would be:
+
+```
+("toystore-app", "1")
+```
+
+**Note**: If rules are not set, it is equivalent to matching all the requests.
+
+### Known limitations
+
+* One HTTPRoute can only be targeted by one rate limit policy.
+* One Gateway can only be targeted by one rate limit policy.
+* Only supporting HTTPRoute/Gateway references from within the same namespace.
+* `hosts` in rules, `spec.rateLimits[].rules`, do not support wildcard prefixes.
+
+## How: Implementation details
+
+### The WASM Filter
+
+On designing kuadrant rate limiting and considering Istio/Envoy's rate limiting offering,
+we hit two limitations
+([described here](https://docs.google.com/document/d/1ve_8ZBq8TK_wnAZHg69M6-f_q1w-mX4vuP1BC1EuEO8/edit#bookmark=id.5wyq2fj56u94)).
```

Comment:
I think this Google Doc is Red Hat internal only.

Reply:
fixed